specification/src/main/asciidoc/webprofile/jpa-cdi/CDI-JPA.adoc

```diff
@@ -0,0 +1,120 @@
+:cdi-spec: https://jakarta.ee/specifications/cdi/4.1
+:jpa-spec: https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1
+
+=== Jakarta Persistence & Jakarta Context and Dependency Injection (CDI) Integration
+This section specifies the requirements for Jakarta EE environments that integrate both {jpa-spec}[Persistence] and {cdi-spec}[CDI].
+
```

Comment:
> I am confused for instance:

@Inject
EntityManager myEntityManager
>
> Which EntityManager is returned? The spec logic seems to imply that this should work and return an EntityManager even though scope and qualifiers were not provided. I think for EntityManagerFactory it is explained more for this case, but I don't know if it is for EntityManager unless I am just missing it.

Wouldn't this be the same as:

```
@PersistenceContext 
EntityManager myEntityManager
```

I think that ^ would just use the first persistence unit found in the application deployment by the Persistence container.  I don't recall any guidance in the Persistence spec on what what happens if there are multiple persistence units defined.

WildFly has a persistence unit hint for marking the default persistence unit (a deployment failure is thrown if there are multiple default persistence units).  I thought we might of discussed this years ago on the jpa spec mailing list but I'm not really sure.

Reply:
> Which `EntityManager` is returned?

The interpretation of such an injection point is specified in great detail by the CDI spec. It's not the place of this text to specify how beans are matched to injection points. All we need to specify here is what are the bean types and qualifiers the bean defined here. Then CDI itself takes care of the rest (according to nontrivial rules).

Reply:
> Wouldn't this be the same as:
>
>     @PersistenceContext 
>     EntityManager myEntityManager
>
> I think that ^ would just use the first persistence unit found in the application deployment by the Persistence container. I don't recall any guidance in the Persistence spec on what what happens if there are multiple persistence units defined.

No. CDI-based injection is never based on "the first thing found". It is always an error if an injection point is ambiguous.

All this is covered in section 5.2 of the CDI 4.1 spec. See, especially, section 5.2.2.

Reply:
Thanks, I will look at [CDI spec section 5.2.2](https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1#unsatisfied_and_ambig_dependencies).

The Persistence Specification rules that would apply to `container-managed persistence contexts` and how they resolve referenced persistence units is https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#a12459

The current thinking (from a brief discussion this week during the Jakarta EE Platform call) is that the [Persistence spec contract for container-managed persistence contexts](https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#container-managed-persistence-contexts) would be followed here as well which includes other Persistence Spec requirements specified through-out the Spec + Javadoc text.  

I think the ask for all of us is to figure out if this Platform change can be included in Jakarta EE 11.

I think that the https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#a12459 section does state how a particular persistence unit could be used if it is the "first thing found" as per `When referencing a persistence unit using the unitName annotation element or persistence-unit-name deployment descriptor element, the visibility scope of the persistence unit is determined by its point of definition` part of the `8.2.2. Persistence Unit Scope` spec text.

Reply:
Look, I think we need to take a step back here and understand that CDI is very powerful and that most of the things we need here are already defined by CDI and we don't need to tie ourselves in knots re-saying things that CDI already says. Sure, for historical reasons the Persistence spec says a bunch of stuff that defines how the legacy `@PersistenceUnit` and `@PersistenceContext` annotations work, but we're trying to move away from that: people have asked for _CDI integration_ and that means defining things in terms of CDI concepts.

The proposed text is:

> the container must make available a bean with ... the scope specified by the _scope_ XML element in _persistence.xml_, or _jakarta.transaction.TransactionScoped_ if no scope is explicitly specified

The semantics of the `@TransactionScoped` annotation already tell us everything we need to know about the default lifecycle of the persistence context: and, yes, the persistence context management is going to be essentially the same as what is defined by section 7.7.2 _Container-managed Transaction-scoped Persistence Context_ of the Persistence spec, but instead of being some special thing connected to the `@PersistenceContext` annotation, it's just something we naturally get for free from CDI. Alternatively, if you explicitly specify a _different_ scope in `persistence.xml`, then CDI controls the lifecycle according to the semantics of that scope.

Similarly, this text:

> the container must make available a bean with bean type _EntityManager_, the qualifiers specified by _qualifier_ XML elements in _persistence.xml_, or _jakarta.enterprise.inject.Default_, if no qualifiers are explicitly specified

completely determines how the persistence unit is matched to an injection point, _according to the rules of CDI_. We don't need to look anywhere else than in the CDI spec to find how that works.

The point is that nobody should be looking in the Persistence spec to find out how CDI injection works, or about the lifecycle of contextual beans in CDI.

Reply:
P.S. An example of how **NOT** to do this stuff is provided by sections 12.4.3 and 12.4.4 of the current JMS spec, which purports to define injection of the `JMSContext` via CDI, but doesn't use the language/concepts of CDI.

Reply:
Oh, perhaps confusion arises from the use of the phrase "container-managed entity manager". This is meant in the generic sense of an entity manager which is created and destroyed by the container. It's not a reference to section 7.7 of the Persistence spec. (Though of course the semantics are *extremely* similar by default.)

Reply:
> The semantics of the @TransactionScoped annotation already tell us everything we need to know about the default lifecycle of the persistence context: and, yes, the persistence context management is going to be essentially the same as what is defined by section 7.7.2 Container-managed Transaction-scoped Persistence Context of the Persistence spec, but instead of being some special thing connected to the @PersistenceContext annotation, it's just something we naturally get for free from CDI. Alternatively, if you explicitly specify a different scope in persistence.xml, then CDI controls the lifecycle according to the semantics of that scope.

At a concrete level, I get that EE deployer code that is already deploying the persistence unit via a call to the PersistenceProvider#createContainerEntityManagerFactory(jakarta.persistence.spi.PersistenceUnitInfo) will meet the additional requirements for CDI in EE environments in a way that integrates with non-CDI code (for the container-managed + transaction-scoped case).  However, I must be wrong as you said that the EE deployer code isn't finding the right persistence unit to use, as that is something we get for free from CDI.

Again, at a concrete level, how does CDI get access to all of the different deployed persistence units that only the EE deployer (internals) has access to?  

Is that made possible by the EE deployer adding something like a https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#declaring_disposer_method (perhaps after jakarta.enterprise.inject.spi.AfterDeploymentValidation is fired?)  

I could see how the deployer_method example linked above could have access to the deployed persistence unit.

Reply:
> However, I must be wrong as you said that the EE deployer code isn't finding the right persistence unit to use, as that is something we get for free from CDI.

CDI doesn't need to know anything about the persistence unit or how it is deployed. All CDI needs is for the container to hand it the definition of a bean.

> Again, at a concrete level, how does CDI get access to all of the different deployed persistence units that only the EE deployer (internals) has access to?

I mean to me this is an implementation detail of the bean that the EE container provides to CDI. We don't need to spell out how this works.

> Is that made possible by the EE deployer adding something like a https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#declaring_disposer_method (perhaps after jakarta.enterprise.inject.spi.AfterDeploymentValidation is fired?)

Yes, exactly, something like that.

It could be:

- a producer method (which creates an `EntityManager`) + disposer method (which calls `close()` on it), 
- a bean definition registered by a portable extension, or even
- something more implementation-specific in the case when the CDI bean manager is integrated more deeply with the rest of the EE container via vendor-defined SPIs.

Of course the Persistence spec still defines how the container interacts with the Persistence provider in order to bootstrap a persistence unit and create an `EntityManagerFactory`. Nothing about that process changes.

The only new thing is that after creating the `EntityManagerFactory`, the container then makes certain beans available to CDI (most likely via some sort of portable extension, but, again, we don't need to specify that).

Reply:
> Again, at a concrete level, how does CDI get access to all of the different deployed persistence units that only the EE deployer (internals) has access to?

Maybe in different wording; CDI does not go looking for deployed persistence units. Instead, whatever code creates a persistence unit would register it with CDI under the scope and qualifiers associated with that persistence units.

The lookup is just normal CDI then.

From a cursory look I would say though that

```
@PersistenceContext 
EntityManager myEntityManager
```

and

```
@Inject
EntityManager myEntityManager
```

should ultimately result in the same entity manager, but I've got to study things a bit more to say for sure.

Reply:
<blockquote>
From a cursory look I would say though that

```
@PersistenceContext 
EntityManager myEntityManager
```

and

```
@Inject
EntityManager myEntityManager
```
should ultimately result in the same entity manager, but I've got to study things a bit more to say for sure.
</blockquote>

Mmmmm, the current wording doesn't guarantee that (though it also doesn't completely disallow it).

One the reason being: `@PrsistenceContext` always injects a transaction-scoped PC. Whereas the new functionality specified here allows PCs with arbitrary CDI scope, as specified in `persistence.xml`.